src/lib/controllers/ui-controller.js

```diff
@@ -651,68 +680,91 @@ export default class UIController {
+    const currentLanguageID = LanguageModelFactory.getLanguageIdFromCode(this.options.items.preferredLanguage.currentValue)
+    this.options.items.lookupLangOverride.setValue(false)
+    this.updateLanguage(currentLanguageID)
+    this.updateVerboseMode()
+    this.updateLemmaTranslations()
+    this.notifyInflectionBrowser()
+
+    this.state.setWatcher('uiActive', this.updateAnnotations.bind(this))
+
+    // Activate listeners
+    MouseDblClick.listen('body', evt => this.getSelectedText(evt))
+    document.addEventListener('keydown', this.handleEscapeKey.bind(this))
+    document.body.addEventListener('Alpheios_Reload', this.handleReload.bind(this))
+    document.body.addEventListener('Alpheios_Embedded_Response', this.disableContent.bind(this))
+    document.body.addEventListener('Alpheios_Page_Load', this.updateAnnotations.bind(this))
+    document.body.addEventListener('Alpheios_Options_Loaded', this.updatePanelOnActivation.bind(this))
+
+    this.isInitialized = true
   }
 
-  static get settingValues () {
-    return {
-      uiTypePanel: 'panel',
-      uiTypePopup: 'popup',
-      verboseMode: 'verbose',
-      enableLemmaTranslations: false
+  /**
+   * Activates a UI controller. If `deactivate()` method unloads some resources, we should restore them here.
+   * @returns {Promise<void>}
+   */
+  async activate () {
+    if (this.isActivated) { return `Already activated` }
+
+    if (!this.isInitialized) { await this.init() }
+
+    // TODO: probably need to set event listeners here
+
+    // Update panel on activation
+    if (this.uiOptions.items.panelOnActivate.currentValue && !this.panel.isOpen()) {
+      this.panel.open()
     }
+    this.isActivated = true
+    this.isDeactivated = false
+    this.state.activate()
   }
 
   /**
-   * Finds a maximal z-index value of elements on a page.
-   * @return {Number}
+   * Deactivates a UI controller. May unload some resources to preserve memory.
+   * In this case an `activate()` method will be responsible for restoring them.
+   * @returns {Promise<void>}
    */
-  getZIndexMax (zIndexDefualt = 2000) {
-    let startTime = new Date().getTime()
-    let zIndex = this.zIndexRecursion(document.querySelector('body'), Number.NEGATIVE_INFINITY)
-    let timeDiff = new Date().getTime() - startTime
-    console.log(`Z-index max value is ${zIndex}, calculation time is ${timeDiff} ms`)
-
-    if (zIndex >= zIndexDefualt) {
-      if (zIndex < Number.POSITIVE_INFINITY) { zIndex++ } // To be one level higher that the highest element on a page
-    } else {
-      zIndex = zIndexDefualt
-    }
+  async deactivate () {
+    if (this.isDeactivated) { return `Already deactivated` }
+
+    // TODO: probably need to remove event listeners here
 
-    return zIndex
+    this.popup.close()
+    this.panel.close()
+    this.isActivated = false
+    this.isDeactivated = true
+    this.state.deactivate()
   }
 
+  // TODO: Do we need to keep it?
+  // reactivate () {
+  //   if (this.state.isDisabled()) {
```

Comment:
I am not sure if the distinction between activate and reactivate is still needed. The check on state.isDisabled does probably need to be handled somewhere, to prevent activation of the webextension on a page with alpheios embedded. But as noted above, this may need to move out of the uicontroller anyway.

Reply:
Agree that it should be moved out of a UI controller